### PR TITLE
Adds Appollo CacheControl capability to graphql-java.

### PR DESCRIPTION
**Pull request by bbakerman**
This is support classes for following this spec

https://github.com/apollographql/apollo-cache-control

### CODE REVIEW

File: src/main/java/graphql/cachecontrol/CacheControl.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        currentExtensions = new LinkedHashMap<>(currentExtensions);
+
+        putHintsInExtensionsMap(currentExtensions);
+
+        return ExecutionResultImpl.newExecutionResult().from((ExecutionResultImpl) executionResult)
+                .extensions(currentExtensions).build();
+    }
+
+    private void putHintsInExtensionsMap(Map<Object, Object> extensions) {
+        List<Map<String, Object>> recordedHints = hints.stream()
```
Comment (andimarek) [inline src/main/java/graphql/cachecontrol/CacheControl.java]: there is a FpKit.map function :-)

File: src/main/java/graphql/cachecontrol/CacheControl.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+     *
+     * @return a new execution result with the hints in the extensions map.
+     */
+    public ExecutionResult addTo(ExecutionResult executionResult) {
+        assertTrue(executionResult instanceof ExecutionResultImpl, "You must pass in an ExecutionResult based on graphql.ExecutionResultImpl");
+        Map<Object, Object> currentExtensions = executionResult.getExtensions();
+        if (currentExtensions == null) {
+            currentExtensions = Collections.emptyMap();
+        }
+        currentExtensions = new LinkedHashMap<>(currentExtensions);
```
Comment (johnliucn925) [inline src/main/java/graphql/cachecontrol/CacheControl.java]: I would move this into putHintsInExtensionsMap and return it like:
   private Map<> putHintsInExtensionsMap(Map<Object, Object> extensions) {
        Map<> currentExtensions = new LinkedHashMap<>(extensions)
         List<Map<String, Object>> recordedHints = map(hints, Hint::toMap);

         Map<String, Object> cacheControl = new LinkedHashMap<>();
        cacheControl.put("version", 1);
        cacheControl.put("hints", recordedHints);

         currentExtensions.put("cacheControl", cacheControl);
        return currentExtensions;
    }

**Comment by kaqqao** [comment]
Why not go for the same approach done for `DataLoaderRegistry`, so that an instance can be registered via `ExecutionInput` instead of having to manually piggy-back it through the context?

**Comment by kaqqao** [comment]
Made a quick PR #1422 in case you like that idea.

**Pull request merged**